cmd/archive/unpack.go

```diff
@@ -36,26 +38,44 @@ const (
 	FailedChroot
 	FailedChdirAfterChroot
 	FailedInvokeUnpack
+	FailedApplyLayer
 )
 
-func main() {
-	ctx := context.Background()
-	op := trace.NewOperation(ctx, "Unpack") // TODO op ID?
-	op.Debugf("New unpack operation created")
-
-	if len(os.Args) != 4 {
-		op.Errorf("Wrong number of arguments passed to unpack binary")
-		os.Exit(WrongArgumentCount)
-	}
-
-	root := os.Args[2]
-
+func customUnpack() {
+	op := setupChroot()
+	op.Debugf("New custom unpack operation created")
 	filterSpec, err := archive.DecodeFilterSpec(op, &os.Args[3])
 	if err != nil {
 		op.Errorf("Couldn't deserialize filterspec %s", os.Args[3])
 		os.Exit(InvalidFilterSpec)
 	}
 
+	if err = archive.UnpackNoChroot(op, os.Stdin, filterSpec, "/"); err != nil {
+		op.Error(err)
+		os.Exit(FailedInvokeUnpack)
+	}
+
+	os.Exit(Success)
+}
+
+func dockerUnpack() {
+	op := setupChroot()
+	op.Debugf("New docker-based unpack operation created")
+	// Untar the archive
+	if _, err := docker.ApplyLayer("/", os.Stdin); err != nil {
+		op.Errorf("Error applying layer: %s", err)
+		os.Exit(FailedApplyLayer)
+	}
+
+	os.Exit(Success)
+
+}
+
+func setupChroot() trace.Operation {
```

Comment:
This function has little to do with the returned trace.Operation.
It even would make unit testing a lot more complicated if log messages are the part of the unit test checks.
Please pass trace.Operation as a function parameter initializing it before setupChroot call.

Reply:
from a conceptual standpoint i see where you're coming from, so i'll refactor this a bit, but 

> It even would make unit testing a lot more complicated if log messages are the part of the unit test checks.

how is unit testing any of this even possible in the first place? it's all disk operations & system calls.

Reply:
Better abstraction and design make things more testable. It is hard to test, not because it works with files and syscalls, it is hard to test because it is designed so. Frequently asking a question "how do I test it?" lead to a better design in the future, that is SOLID and testable.

Reply:
Sure, usually that's correct, but in this case, if I redesign everything against an interface I'd have to mock out all of the function calls, and then I'm just testing a mock. What's the point? The only thing this function does is call system operations & handle errors. I guess I could test the error checking, but if `if err != nil` is a fragile operation, we're in bad shape.